lib/df-detection.c

```diff
@@ -0,0 +1,282 @@
+			report();
+		kfree(current->pairs);
+		current->pairs = NULL;
+		current->df_index = 0;
+		current->df_size = 0;
+	}
+	if (current->addresses != NULL) {
+		current->num_read = 0;
+		current->sz = 0;
+		kfree(current->addresses);
+		current->addresses = NULL;
+	}
+}
+
+void report(void)
+{
+	if (!check_valid_detection())
+		return;
+	int i;
+	unsigned long *first_entries;
+	unsigned int first_nr_entries;
+	unsigned long *second_entries;
+	unsigned int second_nr_entries;
+	unsigned long first_frame = 0;
+	unsigned long second_frame = 0;
+
+	for (i = 0; i < current->df_index; i++) {
+		if (current->addresses[current->pairs[i].first].stack &&
+		    current->addresses[current->pairs[i].second].stack) {
+			first_nr_entries = stack_depot_fetch(
+			    current->addresses[current->pairs[i].first].stack,
+			    &first_entries);
+			second_nr_entries = stack_depot_fetch(
+			    current->addresses[current->pairs[i].second].stack,
+			    &second_entries);
+			int first_index =
+			    filter_stack(first_entries, first_nr_entries);
+			int second_index =
+			    filter_stack(second_entries, second_nr_entries);
+			first_frame = first_entries[first_index];
+			second_frame = second_entries[second_index];
+			char first_buf[64], second_buf[64];
+			int first_len, second_len;
+			/*check if this bug happent in perf_copy_attr*/
+			first_len = scnprintf(first_buf, sizeof(first_buf),
+					      "%ps", (void *)first_frame);
+			second_len = scnprintf(second_buf, sizeof(second_buf),
+					       "%ps", (void *)second_frame);
+			if (strnstr(first_buf, "perf_copy_attr", first_len) &&
+			    strnstr(second_buf, "perf_copy_attr", second_len))
+				continue;
+			pr_err("BUG: multi-read in %ps  "
+			       "between %ps and %ps\n ",
+			       sys_call_table[current->syscall_num],
+			       first_frame, second_frame);
+			pr_err("==============================================="
+			       "===================\n");
+			pr_err("======= First Address Range Stack =======");
+			stack_trace_print(first_entries, first_nr_entries, 0);
+			pr_err("======= Second Address Range Stack =======");
+			stack_trace_print(second_entries, second_nr_entries, 0);
+		} else {
+			pr_err("BUG: Intersection Detected at syscall: %ps\n ",
+			       sys_call_table[current->syscall_num]);
+			pr_err("==============================================="
+			       "===================\n");
+		}
+		pr_err("syscall number %ld  System Call: %pSR\n",
+		       current->syscall_num,
+		       sys_call_table[current->syscall_num]);
+		pr_err(
+		    "First %px len %lu Caller %pSR \nSecond %px len "
+		    "%lu Caller %pSR \n",
+		    current->addresses[current->pairs[i].first].start_address,
+		    current->addresses[current->pairs[i].first].len,
+		    current->addresses[current->pairs[i].first].caller,
+		    current->addresses[current->pairs[i].second].start_address,
+		    current->addresses[current->pairs[i].second].len,
+		    current->addresses[current->pairs[i].second].caller);
```

Comment:
Please add a call to dump_stack_print_info() at the end of the report (after this line).
It will print some extra information about the CPU and process, e.g.:
```
[    2.226036] CPU: 0 PID: 95 Comm: kunit_try_catch Tainted: G        W         5.9.0-rc4-00178-gf5a65530e7de-dirty #1176
[    2.226174] Hardware name: linux,dummy-virt (DT)
```